RFCs/FS-1062-Async-WithCancellation.md

```diff
@@ -0,0 +1,107 @@
+
+# Detailed design
+[design]: #detailed-design
+
+The existing way of approaching this problem in F# is provided here: 
+
+https://gist.github.com/eulerfx/c41d50c6fba8e88cf16a21ed7c3c14bd#file-async-withcancellation-fs
+
+Inlined:
+
+```fsharp
+let withCancellation (ct:CancellationToken) (a:Async<'a>) : Async<'a> = async {
+  let! ct2 = Async.CancellationToken
+  use cts = CancellationTokenSource.CreateLinkedTokenSource (ct, ct2)
+  let tcs = new TaskCompletionSource<'a>()
+  use _reg = cts.Token.Register (fun () -> tcs.TrySetCanceled() |> ignore)
+  let a = async {
+    try
+      let! a = a
+      tcs.TrySetResult a |> ignore
+    with ex ->
+      tcs.TrySetException ex |> ignore }
+  Async.Start (a, cts.Token)
+  return! tcs.Task |> Async.AwaitTask }
+```
+
+Since all methods on `Async` are implemented as `static members`, that pattern should continue to be used with the proposed signature. 
+
+```fsharp
+static member WithCancellation(a : Async<'a>, cancellationToken : CancellationToken) = async {
+  let! ct2 = Async.CancellationToken
+  use cts = CancellationTokenSource.CreateLinkedTokenSource (ct, ct2)
+  let tcs = new TaskCompletionSource<'a>()
+  use _reg = cts.Token.Register (fun () -> tcs.TrySetCanceled() |> ignore)
+  let a = async {
+    try
+      let! a = a
+      tcs.TrySetResult a |> ignore
+    with ex ->
+      tcs.TrySetException ex |> ignore }
+  Async.Start (a, cts.Token)
+  return! tcs.Task |> Async.AwaitTask }
+```
+
+with the same body as shown above.
+
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There is a drawback to adding more members like this, as it does move more towards having a paradox of choice.
+
+# Alternatives
+[alternatives]: #alternatives
+
+We may not want to rely entirely on `TaskCompletionSource<'a>` as it might not be best to context switch between `Async` and `Task`.  It might be worth while exploring creating something like `AsyncCompletionTask<'a>` with a reference implementation in [FSharp.Control.FusionTasks](https://github.com/kekyo/FSharp.Control.FusionTasks/blob/99e3cea2c5121ce00ea6e4c4750103c29a4b586a/FSharp.Control.FusionTasks/Infrastructures.fs#L188).  
+
+# Compatibility
+[compatibility]: #compatibility
+
+Please address all necessary compatibility questions:
+* Is this a breaking change? 
+
+No
+
+* What happens when previous versions of the F# compiler encounter this design addition as source code?
+
+It will not compile.
```

Comment:
In this case, I would expect it to compile, since the change is source and binary compatible (as far as I can tell.)

Reply:
Actually, in this case, I'd say that "N/A" is the appropriate reply for this question, since it's more for new language constructs rather than FSharp.Core constructs.